src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,465 @@ void Frame::denoise(
+            RENDERER_LOG_ERROR("could not load denoiser checkpoint.");
+
+        return result;
     }
 
+    void save_denoiser_checkpoint(
+        const string&                   checkpoint_path,
+        const DenoiserAOV*              denoiser_aov)
+    {
+        // todo: save denoiser checkpoint in the same file.
+        const Deepimf& histograms_image = denoiser_aov->histograms_image();
+        const Deepimf& covariance_image = denoiser_aov->covariance_image();
+        const Deepimf& sum_image = denoiser_aov->sum_image();
+
+        string hist_file_path, cov_file_path, sum_file_path;
+        get_denoiser_checkpoint_paths(checkpoint_path, hist_file_path, cov_file_path, sum_file_path);
+
+        // Add histograms layer.
+        bool result = ImageIO::writeMultiChannelsEXR(histograms_image, hist_file_path.c_str());
+        // Add covariances layer.
+        result = result && ImageIO::writeMultiChannelsEXR(covariance_image, cov_file_path.c_str());
+        // Add sum layer.
+        result = result && ImageIO::writeMultiChannelsEXR(sum_image, sum_file_path.c_str());
+
+        if (!result)
+            RENDERER_LOG_ERROR("could not save denoiser checkpoint.");
+    }
+
+}
+
+bool Frame::load_checkpoint(PermanentShadingResultFrameBufferFactory* buffer_factory)
+{
+    if  (!impl->m_checkpoint_resume)
+        return true;
+
+    bf::path bf_path(impl->m_checkpoint_resume_path.c_str());
+
+    // Check if the file exists.
+    if (!bf::exists(bf_path))
+    {
+        RENDERER_LOG_WARNING("no checkpoint found, starting a new render");
+        return true;
+    }
+
+    // Open the file.
+    GenericProgressiveImageFileReader reader;
+    reader.open(impl->m_checkpoint_resume_path.c_str());
+
+    // First, read layers name and properties.
+    CheckpointProperties checkpoint_props;
+
+    size_t layer_index = 0;
+    while (reader.choose_subimage(layer_index))
+    {
+        CanvasProperties layer_canvas_props;
+        ImageAttributes layer_attributes;
+
+        reader.read_canvas_properties(layer_canvas_props);
+        reader.read_image_attributes(layer_attributes);
+
+        const string layer_name =
+            layer_attributes.exist("name")
+                ? layer_attributes.get<string>("name")
+                : "undefined";
+
+        checkpoint_props.push_back(
+            make_tuple(
+                layer_name,
+                layer_canvas_props,
+                layer_attributes));
+
+        ++layer_index;
+    }
+
+    // Check checkpoint's compatibility.
+    if (!is_checkpoint_compatible(impl->m_checkpoint_resume_path, *this, checkpoint_props))
+    {
+        reader.close();
+        return false;
+    }
+
+    const size_t start_pass = get<2>(checkpoint_props[0]).get<size_t>("appleseed:LastPass") + 1;
+
+    // Check if passes have already been rendered.
+    if (impl->m_pass_count <= start_pass)
+    {
+        RENDERER_LOG_WARNING("passes have previously been rendered and saved into the checkpoint");
```

Comment:
End with a period.

Reply:
In an attempt to make the message clearer, may I suggest:

    the requested passes have already been rendered and saved into the checkpoint.